Src/Plugins/Security/SEC101_043.NuGetCredentialsValidator.cs

```diff
@@ -0,0 +1,154 @@
+            {
+                return ValidationState.NoMatch;
+            }
+
+            if (StaticStringHelper.LikelyPowershellVariable(secret))
+            {
+                return ValidationState.NoMatch;
+            }
+
+            fingerprint = new Fingerprint
+            {
+                Host = host, // technically, this is XML which could contain multiple hosts.
+                Id = id,
+                Secret = secret,
+                Platform = nameof(AssetPlatform.NuGet),
+            };
+
+            return ValidationState.Unknown;
+        }
+
+        protected override ValidationState IsValidDynamicHelper(ref Fingerprint fingerprint, ref string message, ref Dictionary<string, string> options)
+        {
+            string hostXmlAsString = fingerprint.Host;
+            string username = fingerprint.Id;
+            string password = fingerprint.Secret;
+
+            XmlDocument hostXml = new XmlDocument();
+            hostXml.LoadXml(hostXmlAsString);
+
+            // hostXml looks like "<packageSources>...</packageSources>
+            List<string> hosts = hostXml
+                                    ?.ChildNodes[0] // <packageSources>...
+                                        ?.ChildNodes.Cast<XmlNode>().Where(x => x.Name.Equals("add", System.StringComparison.OrdinalIgnoreCase)) // <add ... >  <clear/> (the first node)
+                                            .Select(x => x.Attributes["value"]?.Value ?? x.Attributes["Value"]?.Value).ToList(); // <add key="name of host" value="http://nugetfeedUrl.com" /> (we're looking for URL)
+
+            // Uusually there will probably only be a single host.
+            // In any case, attempt to pass the credentials to each host one by one.
+            using (HttpClient httpClientNoCredentials = CreateHttpClient())
+            using (HttpClient httpClientWithCredentials = CreateHttpClient())
+            {
+                byte[] byteArray = Encoding.ASCII.GetBytes($"{username}:{password}");
+                httpClientWithCredentials.DefaultRequestHeaders.Authorization =
+                    new AuthenticationHeaderValue("Basic", Convert.ToBase64String(byteArray));
+
+                foreach (string host in hosts)
+                {
+                    if (!Uri.IsWellFormedUriString(host, UriKind.Absolute))
+                    {
+                        continue;
+                    }
+
+                    using (HttpResponseMessage responseWithNoCredentials = httpClientNoCredentials
+                    .GetAsync(host, HttpCompletionOption.ResponseHeadersRead)
+                    .GetAwaiter()
+                    .GetResult())
+                    {
+                        switch (responseWithNoCredentials.StatusCode)
+                        {
+                            case HttpStatusCode.OK:
+                                // Credentials not needed, this method of verification is indeterminate.
```

Comment:
>Credentials [](start = 35, length = 11)

this structual choice to instantiate two http clients and embed the actual analysis only in the case where the client w/no credentials is forbidden is an interesting idea but I'd like a cleaner expression of it. You can start by adding braces to all of these switch cases, which is the pattern elsewhere.

Reply:
Done.

---
In reply to: [611239775](https://github.com/microsoft/sarif-pattern-matcher/pull/372#discussion_r611239775) [](ancestors = 611239775)